src/Symfony/Component/Routing/Generator/UrlGenerator.php

```diff
@@ -294,6 +294,15 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
             return $a == $b ? 0 : 1;
         });
 
+        array_walk_recursive($extra, $caster = static function (&$v) use (&$caster) {
+            if ($v instanceof \stdClass) {
+                $v = (array) $v;
+                array_walk_recursive($v, $caster);
+            } elseif (\is_object($v)) {
+                $v = (string) $v;
```

Comment:
if the object does not implement __toString this would throw an exception which is a bc break.
and even if it did, there would be a bc break because http_build_query uses the public properties normally.
see https://3v4l.org/9cDG3

Reply:
> if the object does not implement __toString this would throw an exception which is a bc break.

I added a test case and check for __toString()

> and even if it did, there would be a bc break because http_build_query uses the public properties normally.
see https://3v4l.org/9cDG3

We could check if some public properties exists and do nothing instead of using toString()